src/cmd/keys.js

```diff
@@ -677,5 +610,56 @@ module.exports = class KeysCommand {
 
 		return filename;
 	}
+
+	async getDfuDevice({ deviceID } = {}) {
+		try {
+			let device = await usbUtils.getOneUsbDevice({ idOrName: deviceID, api: this.api, auth: this.auth, ui: this.ui });
+			if (!device.isInDfuMode) {
+				device = await usbUtils.reopenInDfuMode(device);
+			}
+			this._setDfuId(device);
+			return device;
+		} catch (err) {
+			throw new VError(ensureError(err), 'Unable to get DFU device');
+		}
+	}
+
+	_validateSegmentSpecs(segmentName) {
+		let specs = deviceSpecs[this.dfuId] || {};
+		let params = specs[segmentName];
+		let error = null;
+
+		if (!segmentName) {
+			error = "segmentName required. Don't know where to read/write.";
+		} else if (!specs) {
+			error = "dfuId has no specification. Don't know how to read/write.";
+		} else if (!params) {
+			error = `segment ${segmentName} has no specs. Not aware of this segment.`;
+		}
+
+		return { error, specs: params };
+	}
+
+	async _dfuWrite(device, buffer, { altSetting, startAddr, leave, noErase }) {
+		await device.writeOverDfu(buffer, { altSetting, startAddr, leave, noErase });
+	}
+
+	async _dfuRead(device, { altSetting, startAddr, size }) {
+		let buf;
+		try {
+			buf = await device.readOverDfu({ altSetting, startAddr, size });
+		} catch (err) {
+			// FIXME: First time read may fail so we retry
+			buf = await device.readOverDfu({ altSetting, startAddr, size });
+		}
+		return buf;
```

Comment:
Open to improving this logic for more accurate error detection just in case the 'readOverDfu' operation fails again.